### PR TITLE
[SPARK-38306][SQL] Fix ExplainSuite,StatisticsCollectionSuite and StringFunctionsSuite under ANSI mode

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?

Fix ExplainSuite,StatisticsCollectionSuite and StringFunctionsSuite under ANSI mode

### Why are the changes needed?

To set up a new GA test job with ANSI mode on
### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Manually turn on ANSI mode and test

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 1]
```diff
@@ -232,12 +232,11 @@ class ExplainSuite extends ExplainSuiteHelper with DisableAdaptiveExecutionSuite
   }
 
   test("explain for these functions; use range to avoid constant folding") {
-    val df = sql("select ifnull(id, 'x'), nullif(id, 'x'), nvl(id, 'x'), nvl2(id, 'x', 'y') " +
+    val df = sql("select ifnull(id, 1), nullif(id, 1), nvl(id, 1), nvl2(id, 1, 2) " +
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:235]: 'x' can't convert to numeric under ANSI mode

File: sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala [hunk 2]
```diff
@@ -217,8 +217,8 @@ class ExplainSuite extends ExplainSuiteHelper with DisableAdaptiveExecutionSuite
     // AND                                               conjunction
     // OR                                                disjunction
     // ---------------------------------------------------------------------------------------
-    checkKeywordsExistsInExplain(sql("select 'a' || 1 + 2"),
-      "Project [null AS (concat(a, 1) + 2)#x]")
+    checkKeywordsExistsInExplain(sql("select '1' || 1 + 2"),
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:220]: 'a' can't cast to numeric under ANSI mode. This test case is about operator precedence

File: sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala [hunk 3]
```diff
@@ -409,7 +409,7 @@ class StatisticsCollectionSuite extends StatisticsCollectionTestBase with Shared
         val df = spark.range(1000L).select('id,
           'id * 2 as "FLD1",
           'id * 12 as "FLD2",
-          lit("aaa") + 'id as "fld3")
```
Comment (gengliangwang) [inline sql/core/src/test/scala/org/apache/spark/sql/StatisticsCollectionSuite.scala:412]: This results in nulls of double type when ANSI is off.

**Comment by gengliangwang** [comment]
cc @anchovYu as well

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
Merging to master

**Pull request closed without merging**